garden_ai/model_connectors/model_connector.py

```diff
@@ -0,0 +1,366 @@
+        v = v.lower()  # case-insensitive
+        if "url" in values.data and "doi" in values.data:
+            if v == "foundry" and (
+                values.data["url"] is None or values.data["doi"] is None
+            ):
+                raise ValueError(
+                    "For a Foundry repository, both url and doi must be provided"
+                )
+        return v
+
+
+class ModelMetadata(BaseModel):
+    """
+    The ``ModelMetadata`` class represents metadata about an ML model published  \
+    on a public model repository used in an Entrypoint.
+
+    Attributes:
+        model_identifier (str): A short and descriptive name of the model
+        model_repository (ModelRepository): The repository the model is published on.
+        model_version (str): A version identifier
+        datasets (DatasetConnection):
+            One or more dataset records that the model was trained on.
+    """
+
+    model_config = ConfigDict(protected_namespaces=())
+
+    model_identifier: str = Field(...)
+    model_repository: str = Field(...)
+    model_version: Optional[str] = Field(None)
+    datasets: List[DatasetConnection] = Field(default_factory=list)
+
+    @field_validator("model_repository")
+    def must_be_a_supported_repository(cls, model_repository):
+        if model_repository not in [mr.value for mr in ModelRepository]:
+            raise ValueError("is not a supported flavor")
+        return model_repository
+
+
+class ModelConnector(BaseModel, ABC):
+    """Provides attributes and functionality common to models stored in remote git repos.
+
+    Intended to be sub-classed. See `GitHubConnector` and `HFConnector`
+    for example implementations.
+
+    Attributes:
+        repo_url: A str URL to the remote repo. e.g. https://huggingface.co/Garden-AI/sklearn-iris
+        repo_id: A str identifier for the repo in the form 'owner/repo'. e.g. Garden-AI/sklearn-iris
+        branch: str optional git branch to use. Defaults to 'main'.
+        revision: str optional git commit hash to checkout. Defaults to the HEAD of 'main'.
+        local_dir: the local directory to download to repo. Defaults to './models/<repo_name>'
+        enable_imports: enable Python package imports from local_dir.
+        metadata: A `ModelMetadata` object. Will be computed from given information if not provided.
+        readme: an optional readme for the repo.
+        model_dir: base directory for model downloads. defaults to './models'
+    """
+
+    repo_url: Optional[str] = None
+    repo_id: Optional[str] = None
+    branch: str = "main"
+    revision: Optional[str] = None
+    local_dir: Optional[Path] = None
+    enable_imports: bool = True
+    metadata: Optional[ModelMetadata] = None
+    readme: Optional[str] = None
+    model_dir: Optional[Union[Path, str]] = "models"
+
+    model_config = ConfigDict(
+        validate_assignment=True,
+        protected_namespaces=(),
+    )
+
+    def __init__(self, **kwargs):
```

Comment:
Hot take: I think that we might want to express this init logic as a [pydantic model_validator](https://docs.pydantic.dev/latest/concepts/validators/#model-validators). It's not a super common pattern but I think it's warranted here for the following reason: most of these `Optional` attributes aren't really optional at all, because your code here is guaranteeing that they're not None once they've been initialized. 

 For example, if we put your logic around populating `repo_url` from `repo_id` or vice versa into a "before" validator, we could safely assume that both of those attributes exist and are well-formed everywhere else (i.e. in any field-specific validators).